### PR TITLE
Improve GridGenerator::torus, volume case (dim=3)

### LINKED ISSUE #6250: Add more coarse level cells to GridGenerator::truncated_cone
Here is what we get with the current version:

cone-needs-smoothing

We should add more coarse level cells (like in `hyper_ball`) so that the real to unit mapping is not singular.

### PR DESCRIPTION
**Pull request by kronbichler**
Together with @nfehn, I implemented a fix to the torus:

The function GridGenerator::torus() run in 3D (volume mesh) would previously only use a single cell to represent the poloidal shape, which leads to singular mappings similar to how a circle degenerates when meshed with a single cell. The poloidal shape is now represented with 5 cells just like the circle. Furthermore, the function GridGenerator::torus() has gained an optional argument to control the number of cell layers in the toroidal direction. The default manifold set to the torus has also been improved: Now, the TorusManifold is applied on the surface, a CylindricalManifold to the middle cells in toroidal coordinates, and a TransfiniteInterpolationManifold on the cells between the surface and the inner ring. This leads to an excellent mesh quality for all supported scenarios, as opposed to the previous setup that could lead to bad cells in case there were not enough cells in toroidal direction.

Example:
torus_mesh

Closes #6250.

### CODE REVIEW

File: source/grid/grid_generator.cc [hunk 1]
```diff
@@ -725,26 +728,111 @@ namespace GridGenerator
...
+    vertices[0] = p + Point<3>(-1, -1, 0) * (r / std::sqrt(2.0)),
+    vertices[1] = p + Point<3>(+1, -1, 0) * (r / std::sqrt(2.0)),
+    vertices[2] = p + Point<3>(-1, -1, 0) * (r / std::sqrt(2.0) * a),
+    vertices[3] = p + Point<3>(+1, -1, 0) * (r / std::sqrt(2.0) * a),
+    vertices[4] = p + Point<3>(-1, +1, 0) * (r / std::sqrt(2.0) * a),
+    vertices[5] = p + Point<3>(+1, +1, 0) * (r / std::sqrt(2.0) * a),
+    vertices[6] = p + Point<3>(-1, +1, 0) * (r / std::sqrt(2.0)),
+    vertices[7] = p + Point<3>(+1, +1, 0) * (r / std::sqrt(2.0));
+
+    // create remaining vertices by rotating around positive y-axis
```
Comment (nfehn) [inline source/grid/grid_generator.cc]: Could you change the comment to "negative y-axis"? Originally, I used a rotation around positive y-axis but this gives a wrong numbering of vertices (negative cell measure) and then forgot to update the comment. A further explanation "... to ensure positive cell measure" could also be helpful.

Comment (kronbichler) [reply source/grid/grid_generator.cc]: Done!

File: tests/grid/torus_01.output [hunk 2]
```diff
@@ -1,332 +1,916 @@
...
-set arrow from 1.00000, 0.400000, -2.44929e-17 rto 0.269279, 1.91913e-16, 1.25664
-set arrow from 0.300000, -1.71686e-17, 0.519615 rto -0.108828, 0.897598, 0.0628319
-set arrow from -0.500000, -0.400000, 0.866025 rto 0.231841, -5.11243e-16, 1.27396
-set arrow from -1.40000, 1.04983e-32, 1.71451e-16 rto 1.45541e-16, 0.359039, 2.63894
-set arrow from -1.00000, 0.400000, 9.79717e-17 rto -0.269279, 1.91913e-16, 1.88496
-set arrow from -0.300000, -8.08031e-17, -0.519615 rto -1.52359, 0.897598, 0.879646
-set arrow from 0.500000, -0.400000, -0.866025 rto 2.48886, 6.04582e-17, 0.296840
-set arrow from 1.28284, -0.282843, 1.57103e-17 rto 0.476023, 0.476023, 1.61207
-set arrow from 0.717157, 0.282843, -1.75653e-16 rto 0.412553, 0.412553, 0.901206
+set arrow from 0.717157, -0.282843, 0.00000 rto -0.856842, 0.856842, 0.901206
```
Comment (masterleinad) [inline tests/grid/torus_01.output]: `rto` -> `to`

**Comment by drwells** [comment]
This has totally been on my todo list for ages but I never got around to it!

Fixes #6250 (the second half)

File: tests/grid/torus_01.output (hunk 2 above)
Comment (kronbichler) [reply tests/grid/torus_01.output]: Fixed.

**Comment by kronbichler** [comment]
@drwells *ping*

**Comment by drwells** [comment]
This looks good to me.

While you are here, could you add a picture of a cross section (similar to the picture in the original post, but on a coarser grid) to the doxygen page to show the transition between manifolds?

**Comment by kronbichler** [comment]
@drwells I now added a picture of the mesh, colored by manifold ids, on the coarsest level.

**Comment by drwells** [comment]
This looks good. One little thing, though; the image is much wider than the other pictures. Could you scale it down to be about as wide as the cheese grid image?

**Comment by kronbichler** [comment]
You are right, it's too wide - that's what happens on a 4k monitor. If I remember correctly we do not have a mechanism that allows us to easily render figures relative to the html page width via doxygen (without touching style files), so I will simply scale down to 800 pixels width.

**Pull request merged**